### ISSUE TITLE
Run Seq process in `datalust/seq` image as non root user

### ISSUE THREAD

**Issue by msschl** [datalust/seq-tickets#903]
It would be nice to be able to run the seq docker image as an non root user.

**Comment by nblumhardt** [comment]
Hi @msschl - thanks for the note.

Just to clarify, do you mean being able to execute `docker run ... datalust/seq` as a non-root user (without `sudo`)? If so, there's a [`docker` group](https://docs.docker.com/install/linux/linux-postinstall/) that users can be added to on Linux, with some caveats noted in the linked article.

If not, can you please let us know where you're hitting issues? Thanks!

**Comment by msschl** [comment]
No, I mean inside the container itself. With docker you can user USER <uid> in the Dockerfile.

The reason for this is simple: I am running seq as a pod in a kubernetes cluster. This cluster is setup with a default restrictive pod security policy which forces all container to run as an non root user. Therefor the application inside the container must run as an non root user.

**Comment by nblumhardt** [comment]
Thanks for the follow-up; we'll take a look :+1:

**Comment by 101100** [comment]
One of the difficulties currently is that the `Seq` executable itself cannot be run by another user.  Often containers that are designed to run as root can be tricked by mounting volumes with the correct permissions, but the executable itself having the permissions that it does blocks this technique.

**Comment by nblumhardt** [comment]
Hi @101100 - thanks for the note; this sounds like a good first step forwards.

Would adding `x` permission to `Seq` and related executables be enough to unblock this scenario for you? I'm assuming you're using a Kubernetes pod security policy or similar for this?

Investigating now :+1:

**Comment by nblumhardt** [comment]
Ah, and I see there are missing read permissions, too.

Thinking we need:

```
RUN chmod -R a+rX /bin/seq-server
```

**Comment by nblumhardt** [comment]
Looking good - by setting those permissions in the container, I can run it as a non-root user, as long as the mounted `/data` volume has appropriate perms.

**Comment by nblumhardt** [comment]
Have this working smoothly; moved the `chmod` out of the Dockerfile into the host build step, which might save a small amount of space in the image.

Also added:

```
RUN setcap 'cap_net_bind_service=+ep' /bin/seq-server/Seq
```

to allow the (non-root) user to open the container's port 80.

**Comment by 101100** [comment]
@nblumhardt When will this be released?  I'd like to give it a try but `5.1.3364` still doesn't have the executable permissions on `Seq`.

Also, most docker containers expose a port > 1000 on the container so that the `cap_net_bind_service` permission isn't needed since the containers ports are remapped in docker anyway.

**Comment by nblumhardt** [comment]
Thanks for looping back, Jason.

The changes will be released in the next 2020.1 preview image, which is currently looking like early next week. I'll post a note here when it's pushed :+1:

RE changing the container port - I considered making this change but it won't fit the remaining time we have for this release; definitely on the cards for a follow-up in the future, though.

**Comment by becoull** [comment]
Hi everyone, I appreciate this is an old thread. But is it indeed possible to run Seq in a Kubernetes Pod and have the container run as non-root user to align with baseline Pod Security Standards? I am still having issues with this and there seems to be nothing in the Seq docs for K8s deployment. Any information would be really helpful. Thanks.

Pod Security Standards: https://kubernetes.io/docs/concepts/security/pod-security-standards/

@nblumhardt @101100 @msschl

**Comment by 101100** [comment]
I don't make use of Kubernetes in my home server, just `docker-compose`. My desire to run as a non-root user was just due to general security principles. The good news, however, is that I was able to get the container to run as an arbitrary user.

There is a catch, though: the `/data` directory is created as-needed on startup (it isn't in the Seq image) and non-root users don't have permission to create a directory in `/`. To fix this, you need to mount an already-created directory that is owned by the same user as the Seq container.

Just out of curiosity, I tried using a named volume and this also doesn't work as apparently only `root` has permission in volumes by default. I don't know how you'd fix this in Kubernetes, though.

**Comment by nblumhardt** [comment]
Hi @becoull; we're not aware of any current limitations that would prevent what you're trying, but it's also not something we actively track. Opening a new ticket with the details of what's failing would be the best way forward. Cheers!

**Comment by becoull** [comment]
Thanks both for the prompt responses! I work for a company that has a license with Seq and I've been advised to raise a support call. So I'll try this first, and if not I'll create a new ticket here.

All the best :)
